job-server/src/spark.jobserver/AkkaClusterSupervisorActor.scala

```diff
@@ -0,0 +1,217 @@
+                         (ctxName: String, ctxConf: Config, isAdHoc: Boolean)
+                         (successFunc: ActorRef => Unit, failureFunc: Throwable => Unit): Unit = {
+    import akka.pattern.ask
+
+    val resultActor = if (isAdHoc) globalResultActor else context.actorOf(Props(classOf[JobResultActor]))
+    (ref ? JobManagerActor.Initialize(
+      daoActor, Some(resultActor), ctxName, ctxConf, isAdHoc, self))(Timeout(timeoutSecs.second)).onComplete {
+      case Failure(e:Exception) =>
+        logger.info("Failed to send initialize message to context " + ref, e)
+        ref ! PoisonPill
+        failureFunc(e)
+      case Success(JobManagerActor.InitError(t)) =>
+        logger.info("Failed to initialize context " + ref, t)
+        ref ! PoisonPill
+        failureFunc(t)
+      case Success(JobManagerActor.Initialized(_)) =>
+        logger.info("SparkContext {} joined", ctxName)
+        contexts(ctxName) = ref
+        resultActors(ctxName) = resultActor
+        successFunc(ref)
+    }
+
+  }
+
+  private def startContext(name: String, contextConfig: Config, isAdHoc: Boolean)
+                          (successFunc: ActorRef => Unit)(failureFunc: Throwable => Unit) = {
+    require(!(contexts contains name), "There is already a context named " + name)
+    val contextActorName = "jobManager-" + java.util.UUID.randomUUID().toString.substring(16)
+
+    logger.info("Starting context with actor name {} ", contextActorName)
+
+    val pb = Process(s"$managerStartCommand $contextActorName ${selfAddress.toString}")
+    val pio = new ProcessIO(_ => (),
+                        stdout => scala.io.Source.fromInputStream(stdout)
+                          .getLines.foreach(println),
+                        stderr => scala.io.Source.fromInputStream(stderr).getLines().foreach(println))
+    logger.info("Starting to execute sub process {}", pb)
+    val processStart = Try {
+      val process = pb.run(pio)
+      val exitVal = process.exitValue()
+      if (exitVal != 0) {
+        throw new IOException("Failed to launch context process, got exit code " + exitVal)
+      }
+    }
+
+    if (processStart.isSuccess) {
+      contextInitInfos(contextActorName) = (name, contextConfig, isAdHoc)
+      contextCallbacks(contextActorName) = (successFunc, failureFunc)
+    } else {
+      failureFunc(processStart.failed.get)
+    }
+
+  }
+
+  private def addContextsFromConfig(config: Config) {
+    for (contexts <- Try(config.getObject("spark.contexts"))) {
+      contexts.keySet().asScala.foreach { contextName =>
+        val contextConfig = config.getConfig("spark.contexts." + contextName)
+          .withFallback(defaultContextConfig)
+        startContext(contextName, contextConfig, false) { ref =>} {
+          e => logger.error("Unable to start context" + contextName, e)
+        }
+        Thread sleep 1000 // Give some spacing so multiple contexts can be created
```

Comment:
Oh yeah, good catch.  That was old and copied from the days when Spark used global JVM settings to initialize each context and the only way to launch multiple contexts safely was to wait and then clear out the JVM global settings.  Ugh.